sdk/core/core/src/az_span.c

```diff
@@ -643,39 +643,77 @@ AZ_NODISCARD az_result _az_span_url_encode(az_span destination, az_span source,
   int32_t const source_size = az_span_size(source);
   _az_PRECONDITION_VALID_SPAN(destination, source_size, false);
 
-  int32_t result_size = 0;
-  for (int32_t i = 0; i < source_size; ++i)
+  if (source_size == 0)
   {
-    result_size += _az_span_url_should_encode(az_span_ptr(source)[i]) ? 3 : 1;
+    *out_length = 0;
+    return AZ_OK;
   }
 
-  if (az_span_size(destination) < result_size)
+  int32_t const destination_size = az_span_size(destination);
+  if (destination_size < source_size)
   {
     *out_length = 0;
     return AZ_ERROR_INSUFFICIENT_SPAN_SIZE;
   }
 
+  // "Extra space" is measured in units of 2 additional chracters
+  // per single source character ('/' => "%2F").
+  int32_t const extra_space_have = (destination_size - source_size) / 2;
+
+  uint8_t* const dest_begin = az_span_ptr(destination);
+
   uint8_t* const src_ptr = az_span_ptr(source);
-  uint8_t* dest_ptr = az_span_ptr(destination);
+  uint8_t* dest_ptr = dest_begin;
 
-  for (int32_t src_idx = 0; src_idx < source_size; ++src_idx)
+  if (extra_space_have >= source_size)
   {
-    uint8_t c = src_ptr[src_idx];
-    if (!_az_span_url_should_encode(c))
+    // We know that there's enough space even if every character gets encoded.
+    for (int32_t src_idx = 0; src_idx < source_size; ++src_idx)
     {
-      *dest_ptr = c;
-      dest_ptr += 1;
+      uint8_t c = src_ptr[src_idx];
+      if (!_az_span_url_should_encode(c))
+      {
+        *dest_ptr = c;
+        dest_ptr += 1;
+      }
+      else
+      {
+        dest_ptr[0] = '%';
+        dest_ptr[1] = _az_number_to_upper_hex(c >> 4);
+        dest_ptr[2] = _az_number_to_upper_hex(c & 0x0F);
+        dest_ptr += 3;
+      }
     }
-    else
+  }
+  else
+  {
+    // We may or may not have enough space, given whether the input needs much encoding or not.
+    int32_t extra_space_used = 0;
+    for (int32_t src_idx = 0; src_idx < source_size; ++src_idx)
     {
-      dest_ptr[0] = '%';
-      dest_ptr[1] = _az_number_to_upper_hex(c >> 4);
-      dest_ptr[2] = _az_number_to_upper_hex(c & 0x0F);
-      dest_ptr += 3;
+      uint8_t c = src_ptr[src_idx];
+      if (!_az_span_url_should_encode(c))
+      {
+        *dest_ptr = c;
+        dest_ptr += 1;
+      }
+      else
+      {
+        ++extra_space_used;
+        if (extra_space_used > extra_space_have)
```

Comment:
This code duplication and extra-check-inside-loop is unfortunate.

Why don't we have something like:
1) result size = src.size
2) only if dest < src.size * 3, calculate result size and bail early if dest too small.
3) then, just loop through from i to result size without worrying about extra det length checks or logic duplication

Reply:
I wrote the loops with do-while, so that we don't do the very first index comparison now.
Regarding the rest that you described, we currently do have 2 branches, for `dest >= src * 3` and `dest >= src && dest < src * 3`. In the second case, we encode while there is space available, and check while there's still space available, and only if there isn't enough space to encode, we fail. Basically, by avoidng and checking, we avoid looping through the source twice, plus we avoid overrunning the destination in case the source gets changed by another thread after we calculated the size neccessary (although corrupting the memory is not _too_ bad option in this case either).